spec.md

```diff
@@ -542,6 +542,172 @@ The following string-encoded integer literals all have the value `-42`:
+by FIRRTL `connect`{.firrtl} statements (see [@sec:connects]).
+Also unlike nodes, the type of a wire must be explicitly declared.
+
+Example:
+
+``` firrtl
+    wire mywire: UInt<1>
+
+    connect mywire, UInt<1>(0)
+```
+
+The type of a wire is given after the colon (`:`{.firrtl}).
+
+
+## Registers
+
+Registers are stateful elements of a design.
+
+Registers can be connected (see [@sec:connects]).
+The state of the register is controlled through what is connected to it.
+
+The state of a register may be any non-`const`{.firrtl} passive type (see [@sec:passive-types]).
+Registers are always associated with a clock.
+Optionally, registers may have a reset signal.
+
+On every cycle, a register will drive its current value
+and then latch the value it will take on for the next cycle.
+
+The `regreset` keyword is used to declare a register with a reset.
+The `reg` keyword is used to declare a register without a reset.
+
+Examples:
+
+``` firrtl
+wire myclock : Clock
+reg myreg : SInt, myclock
+```
+
+``` firrtl
+wire myclock : Clock
+reg myreg : SInt, myclock
+```
+```firrtl
+wire myclock : Clock
+wire myreset : UInt<1>
+wire myinit : SInt
+regreset myreg : SInt, myclock, myreset, myinit
+```
+
+For both variants of register, the type is given after the colon (`:`{.firrtl}).
+
+Semantically, registers become flip-flops in the design.
+The the next value is latched on the positive edge of the clock.
+For registers without a reset, the initial value is indeterminate (see [@sec:indeterminate-values]).
+
+
+## Output Ports and Input Ports
+
+The way a module interacts with the outside world is through its ports.
+
+Example:
+
+``` firrtl
+    input myinput : UInt<1>
+    output myinput : SInt<8>
+```
+
+Ports are declared with the keywords `input` and `output`.
+The two keywords only differ in the rules for how they may be connected (see [@sec:connects]).
+
+For both variants of port, the type is given after the colon (`:`{.firrtl}).
+
+
+## Submodule Instances
+
+A module in FIRRTL is allowed to contain submodules.
+Each submodule instance must be given a name.
+It must also name the FIRRTL module or external module it is instantiating.
+
+Example:
+
+```firrtl
+inst passthrough of Passthrough
+```
+
+This assumes you have a module named `Passthrough` declared elsewhere in your FIRRTL design.
+The keyword `of`{.firrtl} is used instead of the colon (`:`{.firrtl})
+since `Passthrough` is not a type.
+
+The type of a submodule instance is bundle type determined by its ports.
+Each port creates a field in the bundle of the same name.
+Among these fields, `output` ports are flipped, while `input` fields are unflipped.
+
+For example:
+
+```firrtl
+module Passthrough :
+  input in : UInt<8>
+  output out : UInt<8>
+  connect out, in
+```
+
+The type of the submodule instance `passthrough` above is thus:
+
+```firrtl
+{ flip in : UInt<8>, out : UInt<8> }
+```
+
+
+## Memories
+
+Memories are large stateful elements of a design.
```

Comment:
Nit as with registers: memories may be small, too. They're yet-another stateful element.